semgrep-core/src/core/ast/Ugly_print_AST.ml

```diff
@@ -107,12 +137,49 @@ class base_printer : printer_t =
       (* TODO Handle other kinds *)
       | _ -> print_fail ()
 
-    method print_expr { e; _ } = self#print_expr_kind e
+    method print_expr e =
+      let/ res = self#print_expr_without_parens e in
+      (* An alternative would be to simply have the language-specific printers
+       * add parentheses when they implement `print_opcall` and any other
+       * expressions that need parentheses. Doing it here instead has the
+       * advantage that subclasses need only concern themselves with correctly
+       * printing a given node in isolation, and don't need to worry about
+       * whether it needs to be wrapped in parentheses.
+       *
+       * Otherwise, this logic would need to be duplicated in all of the
+       * language printers as well as the hybrid printer used for autofix. There
+       * will surely be some language-specific differences in parenthesis
+       * insertion, but there is enough commonality between languages that much
+       * of the logic should be reusable. *)
+      Ok (self#add_parens_if_needed (G.E e) res)
+
+    method private print_expr_without_parens { e; _ } = self#print_expr_kind e
     method print_expr_kind _ = print_fail ()
     method print_argument _ = print_fail ()
     method print_arguments _ = print_fail ()
 
+    method print_call e args =
+      match e.G.e with
+      (* Binary operator expressions are desugared to calls, but are different
+       * enough that they merit a separate method here. *)
+      | G.IdSpecial (G.Op _, _) -> self#print_opcall e args
+      (* TODO also handle other kinds of IdSpecial *)
+      | _ -> self#print_ordinary_call e args
+
+    method print_opcall _ _ = print_fail ()
+    method print_ordinary_call _ _ = print_fail ()
+
     (* TODO Add more nodes as needed. *)
+
+    method private needs_parens =
+      function
+      | G.E { e = G.Call ({ e = G.IdSpecial (G.Op _, _); _ }, _); _ } -> true
```

Comment:
maybe add a comment saying that by default we are overly defensive and print extra parents each time.